Change PredictedZeroCrossing algorithm.

markscheel opened the pull request:
## Proposed changes

The new algorithm (the same one as in SpEC), assumes noisy data.
If it cannot be determined (within the error bars of the linear fit) whether the zero crossing
is in the past or in the future, then it returns zero as the crossing time.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

markscheel committed:
PredictedZeroCrossing: use new algorithm.

The new algorithm is the same one as in SpEC, and uses
error bars of the fit to reject cases when noisy data
spuriously indicate a zero crossing.

markscheel commented:
Note that predicted_zero_crossing_value returns a value of zero to indicate that it cannot determine for sure whether the crossing time is in the past or in the future.

This was discussed in two previous PRs (#4004 and #4097) and I am happy to revisit this now that I think all the code that uses it is in PRs.  Other suggestions were for this case to return +infinity , or -infinity, or std::nullopt.

wthrowe commented:
Is the difference between "don't know if we are approaching zero" and "not approaching zero" worth distinguishing?

markscheel commented:
@wthrowe good question.  Currently we do distinguish those cases (same as SpEC):
 * Return positive if we are sure (within error bars) a zero crossing occurs in the future.
 * Return negative if we are sure (within error bars) a zero crossing occurs in the past.
 * Return zero otherwise.
 
However, in practice I think the ZeroCrossingPredictor is currently used in a way so that it cares only about the positive case (e.g. the size control system takes action if it sees that the characteristic speeds are about to cross zero in the near future) and treats the other two cases as equivalent.  (I think; I need to doublecheck the extra states that are not in #4097 yet)

So we could treat the "negative" and "undetermined" cases as equivalent and return +infinity for both, which would simplify some of the logic in size control because the returned value would be guaranteed positive.

My only hesitation: will we ever want to modify the control system logic so that the "negative" and "undetermined" cases result in different behavior?   Maybe.  I could imagine the "negative" case being interpreted as "char speeds are definitely increasing, so it is really safe to transition to a new state", but the transition doesn't happen for the "undetermined" case.

wthrowe commented:
I was confused and thought they were not currently being distinguished.  I agree that knowing which of those two is the case could be useful.

wthrowe commented on src/NumericalAlgorithms/Interpolation/PredictedZeroCrossing.cpp:
```diff
@@ -6,33 +6,56 @@
+
+  // See the doxygen comments for details of the implementation.
+  // For now, set all the x's to zero if denominators are zero;
+  // the actual return type in this case will be determined below.
```
I don't understand this comment.  I assume "return type" is supposed to be something else.

markscheel replied:
I removed the last part of the comment (which didn't make sense) and squashed.

wthrowe reviewed (changes requested):
You can squash this change directly.

wthrowe commented:
Looks good.  Not approving only because this is still dependent.

markscheel commented:
Rebased.

markscheel commented:
Added test_predicted_zero_crossing_indeterminate() in the test to make codecov happy, and squashed.  (It wasn't testing the case when the error bars were large)

kidder commented on src/NumericalAlgorithms/Interpolation/PredictedZeroCrossing.hpp:
```diff
@@ -28,7 +52,9 @@ double predicted_zero_crossing_value(const std::vector<double>& x_values,
  * y_values. The Nth point in the returned DataVector is the value of x for
  * which the linear fit to the points (x[:], y[:][N]) [using python notation]
- * crosses y = 0. The x_values contain M points in the independent variable,
+ * crosses y = 0 (but if we cannot determine within the error bars of the
```
I found the documentation of this function to be repetitive (and therefore confusing...).   Can't it be simplified to point to the other function, just stating that it calls it for each index of the DataVectors?

markscheel replied:
Ok, I will try to simplify this documentation.

(The reason the documentation is long is that there was a lot of confusion in #4004 as to the sizes of the arrays and sizes of the DataVectors)

kidder commented on src/NumericalAlgorithms/Interpolation/PredictedZeroCrossing.hpp:
```diff
@@ -15,6 +15,30 @@ namespace intrp {
+ * are either at x>0 (i.e. in the future if x is time) or at x<0
+ * (i.e. in the past if x is time).  Otherwise we conclude that we
+ * cannot determine even the sign of the crossing value, so we return
+ * zero.
```
why this choice?  why not return the best_fit and the bounds and let the calling code decide what to do?  I'm confused as to why x=0 is special.  Are there constraints on the input values of x?

markscheel replied:
Why x=0 is special: In actual usage, x=0 corresponds to the current time, and all the other input values of x will be negative (i.e. past times).  The y values are things that the control system needs to keep positive, like the min char speeds on the excision bdry, or the difference between the horizon and the excision boundary. So a crossing time with x>0 is in the future (and so the control system probably needs to take action), and a crossing time with x<0 is in the past (so the control system can be assured that it is safe).

And yes, if you want, this logic can be moved to ZeroCrossingPredictor (#4147) thus keeping PredictedZeroCrossing more general. Should I do that?

kidder replied:
I would keep PredictedZeroCrossing more general if it is not an implementation detail (in which case it should be put in a detail namespace).

markscheel commented:
Simplified documentation.

Changed the return type of `predicted_zero_crossing_value` to be a struct that holds bounds and the best-fit value.
(Note that the new version has more memory allocations than the old, less-general version, because the old version kept and returned less information).

wthrowe commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_PredictedZeroCrossing.cpp:
```diff
@@ -47,21 +48,38 @@ void test_predicted_zero_crossing_datavector() {
   std::vector<double> y_values{1.0, 2.1, 1.0, 2.0};
-  CHECK(intrp::predicted_zero_crossing_value(x_values, y_values) == 0.0);
+  const auto zero_crossing =
+      intrp::predicted_zero_crossing_value(x_values, y_values);
```
```
  zero_crossing.lower_bound := -3.65971046088218799
  zero_crossing.best_fit := -6.52631578947368318
  zero_crossing.upper_bound := 15.80900282881401964
```
You can perhaps argue that this result isn't wrong, but the meaning is certainly unintuitive.  (Look at the ordering.)  At minimum you need some explanation in the documentation.  This may not be a better return type than the old one.

markscheel replied:
Oh, wow. I didn't expect this either.  And I think the bounds are actually wrong.

But I think I know what is going on: currently I return the minimum of (x0,x1,x2,x3) as lower_bound and the maximum of (x0,x1,x2,x3) as upper_bound.  But the error bars are so large that (slope+delta_slope) has a different sign than (slope-delta_slope).   So let's consider x0,x1,x2, and x3 to be functions of delta_slope and delta_intercept, so that x0=x1=x2=x3=x_best_fit when delta_slope=delta_intercept=0.  Then as delta_slope varies from zero to its actual value, at least two of the functions x0,x1,x2,x3 will pass through +/- infinity and switch sign.  When this happens, either lower_bound or upper_bound should really be +/- infinity instead of what I return now.

Possible solutions:

1. Revert back to the previous return type and the previous logic (which is all that is needed for the only current use case).
1. Work out the logic to correctly compute lower_bound and upper_bound, and in some cases return lower_bound=-infinity and/or upper_bound=+infinity.  (intercept +/- delta_intercept) will also change sign, so this logic will probably require multiple cases. Note that the code that calls PredictedZeroCrossing (which now checks that lower_bound*upper_bound>0 as a test that a zero-crossing is well determined) would also need to check if the bounds are finite.

Thoughts?

kidder replied:
Okay this seems to be not worth the effort currently.  So reverting to the previous return type is fine, but please add documentation (and an ASSERT?) that the input x_values are supposed to be negative times before the current time zero (i.e. basically the comment you wrote above two days ago) so no one tries to use this function to do something it can't handle correctly

markscheel commented:
Reverted to previous behavior.  Modified docs to explain that x=0 is special. Added ASSERT that x<=0.0.

kidder reviewed (commented):
you can squash if @wthrowe gives the okay

wthrowe commented:
Looks good.  Squash.

markscheel commented:
Squashed.
